base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -785,15 +786,99 @@ public void validateRevocation(ACMEAccount account, ACMERevocation revocation) t
             // No need to check order status since it's guaranteed to be valid.
             // No need to check order expiration since it's irrelevant for revocation.
 
-            logger.info("Account authorized to revoke certificate");
+            logger.info("Account issued the certificate");
             return;
         }
 
         logger.info("Order not found");
 
-        // TODO: Case 2: validate using cert identifiers authorizations
+        // TODO: Case 2: validate using authorization records (if they still exist)
+
+        logger.info("Getting certificate identifiers");
+        Collection<ACMEIdentifier> identifiers = getCertIdentifiers(cert);
+
+        logger.info("Getting account authorizations");
+        Date now = new Date();
+
+        for (ACMEAuthorization authz : database.getAuthorizationsByAccount(account.getID())) {
+
+            String status = authz.getStatus();
+            if (!"valid".equals(status)) {
+                logger.info("Authorization " + authz.getID() + " is " + status);
+                continue;
+            }
+
+            Date expirationTime = authz.getExpirationTime();
+            if (!expirationTime.after(now)) {
+                logger.info("Authorization " + authz.getID() + " has expired");
+                continue;
+            }
+
+            ACMEIdentifier identifier = authz.getIdentifier();
+            logger.info("Authorization " + authz.getID() + " is valid for " + identifier);
+
```

Comment:
I think that processing all authorizations for the account is too expensive if the account has many authorizations (not an unlikely scenario, e.g. web servers that automatically acquire certs for all vhosts).  So I think we need to push the identifier, status and expiry into the database method.  (This implies one query per identifier.  I think that's OK, and better than processing all authorizations for an account).  In the pathological case, e.g. for `InMemoryDatabase`, the filtering can be done inside the database method.  So I am thinking something like:

```java
public Collection<ACMEAuthorization> getValidAuthorizationsForAccountAndIdentifier(
    String accountID, ACMEIdentifier id)
```

or even:

```java
public boolean hasValidAuthorizationForAccountAndIdentifier(
    String accountID, ACMEIdentifier id)
```

We could possibly lift the status and expiry into explicit query parameters; I don't mind either way but probably prefer a more use-case driven API, and we can generalise the method (or add a more general method) later if the need arises.

Reply:
In the new patch I tried adding a `hasValidAuthorizationForAccountAndIdentifier()`, but as you can see, we'd have to implement the following logics in all databases and they have to be kept consistent:
* checking that authorization's status is valid
* checking authorization's expiration time against the provided time
* matching authorization's identifier against the provided identifier (which may use a wildcard)

The method takes a `time` parameter to make sure all authorization's expiration time are compared against the same time.

Implementing wildcard matching at database-level will be very database-specific, so ideally we should run tests for each database to make sure that the implementation for that database is correct.

On the other hand, `getAuthorizationsByAccount()` is very generic and straight forward to implement. The wildcard can be implemented consistently with Java regex, so we don't have to test individual database implementations. The method might run a bit slow if the account has many valid & non-expired authorizations, but it really depends on the server's auto-purging policy.

Another option is, if we have the `certs` table (which could expire as well) as we discussed in the email, we can determine the cert ownership easily without using the authorization records. So even if `getAuthorizationsByAccount()` is slow, it's most likely won't be used.

I think the only reliable revocation method is using the private key. Revocatoin using account authorization relies on whether the server still has the orders, authorizations, or the certs, all of which may disappear depending on the server's auto-purging policy.

@cipherboy What do you think?

Reply:
I've updated the PR to support 2 methods:
* getRevocationAuthorizations()
* hasRevocationAuthorization()

They both will look for valid and non-expired authorizations. The first one will return all authorizations owned by the account, but the second one will only check the authorization for a specific identifier. The first one might be easier to implement so it should be the default (i.e. fallback) implementation. The second one might be more capable to handle large data, but might be more difficult to implement the wildcard matching for all identifier types, so it will be optional.

Reply:
I'm sorry I'm confused why this is a problem

> matching authorization's identifier against the provided identifier (which may use a wildcard)
> Implementing wildcard matching at database-level will be very database-specific, so ideally we should run tests for each database to make sure that the implementation for that database is correct.

As I understand it, there's one ACMEIdentifier type: DNS. There's such thing as a "wildcard DNS" entry on certs, (such as `*.redhat.com`). We aren't looking to match _any_ `redhat.com` entry in the database, we're literally looking for the account with authorization to all off `*.redhat.com`.

Wouldn't that be just another string match? Why would that require a special handling? Presumably you mean globbing or some other wild-card / greedy matching in the database itself.

Reply:
I'm not familiar yet with how the wildcard is supposed to work. Let me try to add enrollment with wildcard first, then we'll come back to revocation. Thanks!

Reply:
I don't think wildcards complicate the situation that much.

If a DNS identifier does not have a wildcard, you look for a (non-wildcard) authorization for that identifier.

If a DNS identifier does have a wildcard `"*.{base}"`, you look for a wildcard=true authorization for `{base}`.

A wildcard authorization is an authorization to issue a certificate with identifier `*.{base}`.  It is not an authorization for a certificate with `{label}.{base}` for some label `label`.  So for each name on the cert to be revoked we only need to look for one kind of authorization object.

In practice, what this means as the the ACMEDatabase API has to change just a little bit:

```java
hasRevocationAuthorization(String accountId, DateTime now, ACMEIdentifier ident, boolean wildcard)
```

with minor flow on effects e.g. (pseudocode):
```java
boolean wildcard = false;
if (ident.isWildcard()) {
  ident.stripWildcard();
  wildcard = false;
}
```

I'm happy for this to be done now, or in a later refactoring.

Reply:
Currently the wildcard info is stored in `ACMEAuthorization` instead of `ACMEIdentifier` but let's revisit them when we implement wildcard.